src/core/geometry/qgslinestring.cpp

```diff
@@ -620,6 +620,125 @@ QPolygonF QgsLineString::asQPolygonF() const
   return points;
 }
 
+
+void simplifySection( int i, int j, const double *x, const double *y, std::vector< bool > &usePoint, const double distanceToleranceSquared, const double epsilon )
+{
+  if ( i + 1 == j )
+  {
+    return;
+  }
+
+  double maxDistanceSquared = -1.0;
+
+  int maxIndex = i;
+  double mx, my;
+
+  for ( int k = i + 1; k < j; k++ )
+  {
+    const double distanceSquared = QgsGeometryUtilsBase::sqrDistToLine(
+                                     x[k], y[k], x[i], y[i], x[j], y[j], mx, my, epsilon );
+
+    if ( distanceSquared > maxDistanceSquared )
+    {
+      maxDistanceSquared = distanceSquared;
+      maxIndex = k;
+    }
+  }
+  if ( maxDistanceSquared <= distanceToleranceSquared )
+  {
+    for ( int k = i + 1; k < j; k++ )
+    {
+      usePoint[k] = false;
+    }
+  }
+  else
+  {
+    simplifySection( i, maxIndex, x, y, usePoint, distanceToleranceSquared, epsilon );
+    simplifySection( maxIndex, j, x, y, usePoint, distanceToleranceSquared, epsilon );
+  }
+};
+
+QgsLineString *QgsLineString::simplifyByDistance( double tolerance ) const
+{
+  if ( mX.empty() )
+  {
+    return new QgsLineString();
+  }
+
+  // ported from GEOS DouglasPeuckerLineSimplifier::simplify
```

Comment:
Did you measure it? It looks like a dangerous precedent if we need to directly copy/paste geos code any time we need performance, mainly because It adds up code maintenance on something we should just rely on.

Reply:
Yes, of course! I'm not a fan of doing work just for the sake of it! 😝I've spent literally weeks of time staring at profiles of the QgsGeometry classes and I'm well aware of the bottlenecks in those classes.